cassiopeia/data/CassiopeiaTree.py

```diff
@@ -0,0 +1,357 @@
+    for working with lineage tracing experiments. At its core, it stores
+    three main items - a tree, a character matrix, and meta data associated
+    with the data.
+
+    The tree can be fed into the object via Ete3, Networkx, or can be inferred
+    using one of the CassiopeiaSolver algorithms in the `solver` module.
+
+    A character matrix can be stored in the object, containing the states
+    observed for each cell. In typical lineage tracing experiments, these are
+    integer representations of the indels observed at each unique cut site.
+
+    Meta data for cells or characters can also be stored in this object. These
+    items can be categorical or numerical in nature. Common examples of cell
+    meta data are the cluster identity, tissue identity, or number of target-site
+    UMIs per cell. These items can be used in downstream analyses, for example
+    the FitchCount algorithm which infers the number of transitions between
+    categorical variables (e.g., tissues). Common examples of character meta
+    data are the proportion of missing data for each character or the entropy
+    of states. These are good statistics to have for feature selection.
+
+    TODO(mattjones315): Add experimental meta data as arguments.
+    TODO(mattjones315): Add functionality that mutates the underlying tree
+        structure: pruning lineages, collapsing mutationless edges, and
+        collapsing unifurcations
+    TODO(mattjones315): Add utility methods to compute the colless index
+        and the cophenetic correlation wrt to some cell meta item
+
+    Args:
+        character_matrix: The character matrix for the lineage.
+        cell_meta: Per-cell meta data
+        character_meta: Per-character meta data
+        priors: A dictionary storing the probability of a character mutating
+            to a particular state.
+        tree: A tree for the lineage. 
+    """
+
+    def __init__(
+        self,
+        character_matrix: pd.DataFrame,
+        cell_meta: Optional[pd.DataFrame] = None,
+        character_meta: Optional[pd.DataFrame] = None,
+        priors: Optional[Dict[int, Dict[int, float]]] = None,
+        tree: Optional[Union[str, ete3.Tree, nx.DiGraph]] = None
+    ):
+
+        self.character_matrix = character_matrix
+        self.cell_meta = cell_meta
+        self.character_meta = character_meta
+        self.priors = priors
+        self.__network = None
+
+        self.populate_tree(tree)
+
+    def populate_tree(self, tree: Union[str, ete3.Tree, nx.DiGraph]):
+
+        if isinstance(tree, nx.DiGraph):
+            self.__network = tree
+        if isinstance(tree, str):
+            self.__network = utilities.newick_to_networkx(tree)
+        if isinstance(tree, ete3.Tree):
+            self.__network = utilities.ete3_to_networkx(tree)
+
+        if self.__network is not None:
+
+            # add character states
+            for n in self.nodes:
+                if n in self.character_matrix.index.values:
```

Comment:
I understand where you coming from on this - and with previous comments I think I've addressed why I don't think the second option is an ideal way of doing things.

As for passing in the character matrix to `populate_tree` (or for that matter, a dictionary mapping leaves to states), this essentially boils down to the issue of what we expect to be in the `CassiopeiaTree`. Realistically, what would not having the character matrix buy us? I admit it would grant us a bit of flexibility on the initialization of the `CassiopeiaTree`, but in the end most functionality would require us to specify the character matrix (tree inference, ancestral character state reconstruction, branch length estimation, etc.).